KAFKA-4827: Correctly encode special chars while creating URI objects

wicknicks opened the pull request:
Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
Correctly encode special chars while creating URI objects

Signed-off-by: Arjun Satish <arjun@confluent.io>

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

wicknicks commented:
retest this please

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
SUCCESS 
 8069 tests run, 5 skipped, 0 failed.
--none--

asfgit commented:
SUCCESS 
 8069 tests run, 5 skipped, 0 failed.
--none--

asfgit commented:
SUCCESS 
 8069 tests run, 5 skipped, 0 failed.
--none--

wicknicks commented:
@kkonstantine and @rhauch: Could you please review the changes in this PR. Thanks!

soenkeliebau commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:
```diff
@@ -99,7 +100,7 @@ public Response createConnector(final @QueryParam("forward") Boolean forward,
         Herder.Created<ConnectorInfo> info = completeOrForwardRequest(cb, "/connectors", "POST", createRequest,
                 new TypeReference<ConnectorInfo>() { }, new CreatedConnectorInfoTranslator(), forward);
-        return Response.created(URI.create("/connectors/" + name)).entity(info.result()).build();
+        return Response.created(URI.create("/connectors/" + URLEncoder.encode(name, "UTF-8"))).entity(info.result()).build();
```
I am not sure if URLEncoder will behave correctly in this case ([see this SO](https://stackoverflow.com/questions/14321873/java-url-encoding-urlencoder-vs-uri)). The one large difference to the URI object is, that it will encode spaces as + which I think will be interpreted as an actual plus sign by connect (tested very briefly, may need further investigation with your actual changes). The correct encoding would be %20 - since I believe we want to allow spaces in the middle of connector names. I think using the URI object for encoding here would be a better choice.

wicknicks replied:
Thanks for the comment, @soenkeliebau. Regarding the '/', there is a check in createConnector which throws a BadRequestException if there is a '/' in the name of the connector. Hence, we should not encounter a scenario where a '/' needs to be encoded in the name. 

Characters like '@' are not percent encoded because they don't have a special meaning (yet) even though they are reserved (the URIBuilders probably skip such characters from being percent encoded to keep the URL short to better support HTTP clients which have lower limits on the length of the URI than others).

Let me know what you think.

wicknicks replied:
Also, I am going to switch to use javax.ws.rs.core.UriBuilder to construct the Location URI. It seems better with handling special characters.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:
```diff
@@ -99,7 +100,7 @@ public Response createConnector(final @QueryParam("forward") Boolean forward,
         Herder.Created<ConnectorInfo> info = completeOrForwardRequest(cb, "/connectors", "POST", createRequest,
                 new TypeReference<ConnectorInfo>() { }, new CreatedConnectorInfoTranslator(), forward);
-        return Response.created(URI.create("/connectors/" + name)).entity(info.result()).build();
+        return Response.created(URI.create("/connectors/" + URLEncoder.encode(name, "UTF-8"))).entity(info.result()).build();
```
This line is getting a little busy. You might consider pulling the encoding logic into a separate method (especially because addressing @soenkeliebau's concerns might result in more code to do this) and then adding a local variable for the encoded name. This same method could be used on line 149, too.

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResourceTest.java:
```diff
@@ -65,8 +67,15 @@
     // URL construction properly, avoiding //, which will mess up routing in the REST server
     private static final String LEADER_URL = "http://leader:8083/";
     private static final String CONNECTOR_NAME = "test";
+    private static final String CONNECTOR_NAME_SPECIAL_CHARS = "test?a=b&c=d\rx=1.1\n><";
```
Can you add a few more characters, like space, tab, `%`, `+`, etc.? Some of those are used in the URL encoding, so including them would help ensure we're properly encoding and decoding.

rhauch reviewed (commented):
Thanks, @wicknicks. Nice job - just a few comments.

wicknicks committed:
Replace URLEncoder with URIBuilder

Also, update the tests to pass some additional characters in the connector
name along with adding a decode step using the URI class.

Signed-off-by: Arjun Satish <arjun@confluent.io>

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 7961 tests run, 5 skipped, 1 failed.
--none--

wicknicks commented:
retest this please

asfgit commented:
SUCCESS 
 8069 tests run, 5 skipped, 0 failed.
--none--

asfgit commented:
SUCCESS 
 8069 tests run, 5 skipped, 0 failed.
--none--

asfgit commented:
SUCCESS 
 8069 tests run, 5 skipped, 0 failed.
--none--

rhauch reviewed (approved):
LGTM. Thanks, @wicknicks!

rhauch commented:
@ewencp can you please review and merge? This should also be backported, presumably as far back as the `0.10.0` branch for any future patch releases.

ewencp reviewed (approved):
LGTM. I can merge and cherry-pick back to 0.11, but `ConnectorType` enum and extended `ConnectorInfo` class containing the type wasn't added until 0.11 so previous branches would need modifications.

ewencp commented:
Actually, I take it back, we can only get it back to 1.0 in its current form because some additions for KIP-151 didn't make it in until 1.0.0

ewencp commented:
Ok, merged to trunk and 1.0. @wicknicks do you want to prepare a follow up PR for the older branches? It's probably straightforward, just need to remove the extra parameters in the tests, but it would be good to get the jenkins feedback.
